src/utils/generatePermalink.js

```diff
@@ -1,29 +1,36 @@
 "use strict";
 
 const config = require("../_data/config.json");
-const getLang = require("./getLang.js");
+const { EleventyI18nPlugin } = require("@11ty/eleventy");
 const TemplateConfig = require("@11ty/eleventy/src/TemplateConfig.js");
-const translations = require("../_data/translations.json");
 
-module.exports = (data, collectionType) => {
+module.exports = (data, collectionType, collectionSlug) => {
     /* If this post is a "stub" with no localized title, we assume it does not exist and prevent it from building. */
-    if (!Object.prototype.hasOwnProperty.call(data, "title")) {
+    if (!data.hasOwnProperty("title")) {
         return false;
     }
 
-    const lang = getLang(data.page.filePathStem, collectionType);
+    const lang = EleventyI18nPlugin.LangUtils.getLanguageCodeFromInputPath(data.page.inputPath);
     const langSlug = config.languages[lang].slug || lang;
     const eleventyConfig = new TemplateConfig();
-    const slugFilter = eleventyConfig.userConfig.getFilter("slugify");
-    const slug = slugFilter(data.title);
+    const slugify = eleventyConfig.userConfig.getFilter("slugify");
 
     if (collectionType === "pages") {
+        /* If the page is a 404 page, return 404.html, optionally prepended with the language code. */
+        if (data.page.fileSlug === "404") {
+            return (lang === data.defaultLanguage) ? "/404.html" : `/${langSlug}/404.html`;
+        }
+
+        /** If the page is the index page, the basepath, optionally prepended with the language code. */
         if (data.page.fileSlug === lang) {
-            return (lang === config.defaultLanguage) ? "/" : `/${langSlug}/`;
+            return (lang === data.defaultLanguage) ? "/" : `/${langSlug}/`;
         }
 
-        return (lang === config.defaultLanguage) ? `/${slug}/` : `/${langSlug}/${slug}/`;
+        /* If the page is not the index page, return the slugified page title, optionally prepended with the language code. */
+        const slug = slugify(data.title);
+        return (lang === data.defaultLanguage) ? `/${slug}/` : `/${langSlug}/${slug}/${data.hasOwnProperty("pagination") && data.pagination.pageNumber > 0 ? "page/" + data.pagination.pageNumber + 1 + "/" : ""}`;
```

Comment:
Does this work as you'd expect? I'm wondering because it appears that you are doing both string concatenation and addition; `"page/" + data.pagination.pageNumber + 1 + "/"`.